src/env.cc

```diff
@@ -454,6 +463,73 @@ void Environment::RunAndClearNativeImmediates() {
 }
 
 
+void Environment::ScheduleTimer(int64_t duration_ms) {
+  uv_timer_start(timer_handle(), RunTimers, duration_ms, 0);
+}
+
+void Environment::ToggleTimerRef(bool ref) {
+  if (ref) {
+    uv_ref(reinterpret_cast<uv_handle_t*>(timer_handle()));
+  } else {
+    uv_unref(reinterpret_cast<uv_handle_t*>(timer_handle()));
+  }
+}
+
+void Environment::RunTimers(uv_timer_t* handle) {
+  Environment* env = Environment::from_timer_handle(handle);
+
+  if (!env->can_call_into_js())
+    return;
+
+  HandleScope handle_scope(env->isolate());
+  Context::Scope context_scope(env->context());
+
+  Local<Object> process = env->process_object();
+  InternalCallbackScope scope(env, process, {0, 0});
+
+  Local<Function> cb = env->timers_callback_function();
+  MaybeLocal<Value> ret;
+  Local<Value> arg = env->GetNow();
+  // This code will loop until all currently due timers will process. It is
+  // impossible for us to end up in an infinite loop due to how the JS-side
+  // is structured.
+  do {
+    TryCatch try_catch(env->isolate());
+    try_catch.SetVerbose(true);
+    ret = cb->Call(env->context(), process, 1, &arg);
+  } while (ret.IsEmpty() && env->can_call_into_js());
+
+  if (!env->can_call_into_js())
```

Comment:
Same reply as above but also we need to exit here if the loop broke out because of not being able to call into JS. This is mainly important for @addaleax's workers PR (it should exit on the first condition above when running in the main thread).

Reply:
I’m not sure about this one – we don’t attempt to make any directly calls into JS here below this, right?

> otherwise we would fail on ToLocalChecked below

In that case it would be a better idea to return `if (ret.IsEmpty())` – but then again, do we not want to re-schedule timers if there was a throw? I.e. what happens if `ret` is empty but ont because of execution termination?

Reply:
@addaleax Right, that was my questions yesterday but maybe badly phrased. Is it possible that `can_call_into_js()` goes to `false` and then back to `true`. If it's not at the moment, I would say this is fine as is (with the check being changed to `ret.IsEmpty()`)  and we can cross the bridge later. I don't want to overthinking this for potential future change because this does allow us to remove like 3 boundary crossings from timers.

Reply:
Yes, I think this is fine with `ret.IsEmpty()`.

Reply:
I've added a NOTE that will make the behaviour clear. Let me know if it doesn't make sense?